### PR TITLE
fix bug when some error occurs in loader arbiter's syncBinlog may block

### PR DESCRIPTION
**Pull request by lichunzhu**
### What problem does this PR solve? 
[TOOL-1488](https://internal.pingcap.net/jira/browse/TOOL-1488)
When loader executes some dmls and meet an error in `loader.Run()`, it will shut down, report an error and stop dealing with chan `loader.input`. But `syncBinlogs` in `arbiter/server.go` will keep on pushing items into `loader.input`, when it get full, function `syncBinlogs` will be blocked and then arbiter will be unable to quit or issue the error.
### What is changed and how it works?
Add `syncCtx` and `select` in function `syncBinlogs` to avoid blocking.

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
Modified arbiter will quit as we expect.
image

Code changes

Side effects

Related changes

### CODE REVIEW

File: arbiter/server_test.go [hunk 1]
```diff
@@ -407,7 +407,7 @@ func (s *syncBinlogsSuite) TestShouldSendBinlogToLoader(c *C) {
 	}()
 	ld := dummyLoader{input: dest}
 
-	err := syncBinlogs(source, &ld)
+	err := syncBinlogs(context.Background(), source, &ld)
```
Comment (suzaku) [inline arbiter/server_test.go]: Add a test case here to show that `syncBinlogs` can be canceled from outside?

Comment (lichunzhu) [reply arbiter/server_test.go]: I add another test case in the end

File: arbiter/server.go [hunk 2]
```diff
@@ -182,22 +188,26 @@ func (s *Server) Run() error {
...
+		syncErr = syncBinlogs(syncCtx, s.kafkaReader.Messages(), s.load)
 		if syncErr != nil {
 			s.Close()
 		}
 	}()
 
 	err := s.load.Run()
 	if err != nil {
+		syncCancel()
 		s.Close()
```
Comment (suzaku) [inline arbiter/server.go]: Maybe we can fix this problem in a simpler way? For example, it seems buggy that `Reader.Close` doesn't `close` the `Messages` channel.

Comment (lichunzhu) [reply arbiter/server.go]: `Reader.Close` has closed successfully, but `Reader.msgs` channel has a buffer size of 1024. Func `syncCancel` can still read the buffered `msg` although it is closed.

File: arbiter/server_test.go [hunk 3]
```diff
@@ -418,3 +418,38 @@ func (s *syncBinlogsSuite) TestShouldSendBinlogToLoader(c *C) {
...
+func (s *syncBinlogsSuite) TestShouldQuitWhenSomeErrorOccurs(c *C) {
+	readerMsgs := make(chan *reader.Message, 1024)
+	dummyLoaderImpl := &dummyLoader{
+		successes: make(chan *loader.Txn),
+		// input is set small to trigger blocking easily
+		input: make(chan *loader.Txn, 1),
+	}
+	msg := s.createMsg("test42", "users", "alter table users add column gender smallint")
+	ctx, cancel := context.WithCancel(context.Background())
+	syncCtx, syncCancel := context.WithCancel(ctx)
```
[the remark this answers is not in the document]
Comment (lichunzhu) [reply arbiter/server_test.go]: Revised

**Comment by lichunzhu** [comment]
/run-integration-tests tidb=release-3.0 tikv=release-3.0 pd=release-3.0

File: arbiter/server.go [hunk 4]
```diff
@@ -162,7 +162,13 @@ func (s *Server) Close() error {
 
 // Run runs the Server, will quit once encounter error or Server is closed
 func (s *Server) Run() error {
-	defer s.downDB.Close()
+	defer func() {
+		if s.downDB != nil {
+			s.downDB.Close()
+		} else {
+			log.Error("Invalid downDB address, please check the config")
```
Comment (suzaku) [inline arbiter/server.go]: When would `downDB` be `nil`?

Comment (lichunzhu) [reply arbiter/server.go]: Revised. This part was modified because at first I wanna test the whole `Run` function.

File: arbiter/server.go [hunk 5]
```diff
@@ -182,22 +182,26 @@ func (s *Server) Run() error {
...
 
 	err := s.load.Run()
 	if err != nil {
+		syncCancel()
 		s.Close()
 	}
 
 	wg.Wait()
 
+	// to pass go check
```
Comment (july2993) [inline arbiter/server.go]: address the comment, we don't write code for static check tools... the problem is why we need to cancel here?
just remove the comment is better to add a "to pass go check"

**Comment by lichunzhu** [comment]
@july2993 There is a bug in the previous test. In `readerMsgs <- msg` when `ctx` is cancelled directly there might be no `msg` in channel `readerMsgs` and `syncBinlogs` will get blocked since channel `readerMsgs` is not closed. Now 3 msgs will be sent to `readerMsgs` and after `ctx` is cancelled `readerMsgs` will be closed. PTAL.

**Pull request merged**